File: packages/ui/src/api/socketService.ts
import { io, Socket } from 'socket.io-client';

const soketUrl = import.meta.env.VITE_BASE_URI;

class SocketService {
  private socket: Socket | null = null;

  connect() {
    if (!this.socket) {
      this.socket = io(soketUrl, {
        path: '/ws',
        transports: ['websocket'],
        reconnection: true,
      });

      this.socket.on('connect', () => {
        console.log('Connected to Socket.IO server with ID:', this.socket?.id);
      });

      this.socket.on('connect_error', (error) => {
        console.error('Connection error:', error.message);
      });

      this.socket.on('disconnect', (reason) => {
        console.log('Disconnected from server:', reason);
      });
    } else {
      console.log('Socket already connected.');
    }
  }

  joinRoom(roomId: string) {
    if (this.socket) {
      this.socket.emit('join_room', roomId);
      console.log(`Joined room: ${roomId}`);
    } else {
      console.warn('Socket is not connected. Cannot join room.');
    }
  }

  onNewComment(callback: (newComment: any) => void) {
    this.socket?.on('newComment', callback);
  }

  disconnect() {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
    }
  }

  on(event: string, callback: (data: any) => void) {
    if (this.socket) {
      this.socket.on(event, callback);
    }
  }

  emit(event: string, data: any) {
    if (this.socket) {
      this.socket.emit(event, data);
    }
  }

  onOrderUpdated(callback: (data: any) => void) {
    if (this.socket) {
      this.socket.on('orderUpdated', callback);
    } else {
      console.warn('Socket is not connected. Cannot listen to order updates.');
    }
  }

  get isConnected() {
    return this.socket?.connected || false;
  }
}

export default new SocketService();

File: packages/app/lib/persistance/socket_service.dart
import 'package:socket_io_client/socket_io_client.dart' as IO;
import 'package:tedbook/persistance/remote/api_provider.dart';
import 'package:tedbook/utils/utils.dart';

class SocketService {
  late IO.Socket socket;

  void connect() {
    // Configure the socket connection
    socket = IO.io(
        ApiProvider.baseUrl,
        IO.OptionBuilder()
            .setTransports(['websocket']) // use web socket transport
            .setPath('/ws') // Ensure this matches your server configuration
            .enableAutoConnect() // Auto-connect when the service is initialized
            .build());

    // Handle connection events
    socket.onConnect((_) {
      debugLog('Connected to Socket.IO server with ID: ${socket.id}');
    });

    socket.onDisconnect((_) {
      debugLog('Disconnected from server');
    });

    socket.onConnectError((error) {
      debugLog('Connection error: $error');
    });

    // Listen for new comments from the server
    socket.on('newComment', (data) {
      debugLog('New comment received: $data');
      // Handle the received comment data (you might want to parse it)
    });
  }

  void joinRoom({required String orderId}) {
    socket.emit('join_room', orderId);
    debugLog('Joined room: $orderId');
  }

  void sendMessage({
    required String orderId,
    required String commenterRole,
    required String commentText,
  }) {
    final message = {
      'orderId': orderId,
      'commenterRole': commenterRole,
      'commentText': commentText,
    };
    socket.emit('sendMessage', message);
    debugLog('Message sent: $message');
  }

  void disconnect() {
    socket.disconnect();
  }
}
